org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSContentAssistProcessor.java

```diff
@@ -216,40 +214,28 @@ private String createErrorMessage(int offset, Exception ex) {
 
 	private void initiateLanguageServers(IDocument document) {
 		if (currentDocument != document) {
-			this.currentDocument = document;
-			if (this.completionLanguageServersFuture != null) {
-				try {
-					this.completionLanguageServersFuture.cancel(true);
-				} catch (CancellationException ex) {
-					// nothing
-				}
-			}
-			if (this.contextInformationLanguageServersFuture != null) {
-				try {
-					this.contextInformationLanguageServersFuture.cancel(true);
-				} catch (CancellationException ex) {
-					// nothing
-				}
-			}
-			this.completionTriggerChars = NO_CHARS;
-			this.contextTriggerChars = NO_CHARS;
+			currentDocument = document;
+			cancelFuture(completionLanguageServersFuture);
+			cancelFuture(contextInformationLanguageServersFuture);
 
-			this.completionLanguageServersFuture = LanguageServers.forDocument(document)
+			completionTriggerChars = NO_CHARS;
+			contextTriggerChars = NO_CHARS;
+
+			completionLanguageServersFuture = LanguageServers.forDocument(document)
 					.withFilter(capabilities -> capabilities.getCompletionProvider() != null) //
 					.collectAll((w, ls) -> {
-						CompletionOptions provider = castNonNull(w.getServerCapabilities()).getCompletionProvider();
+						List<String> triggerChars = castNonNull(w.getServerCapabilities()).getCompletionProvider().getTriggerCharacters();
 						synchronized (completionTriggerCharsSemaphore) {
-							completionTriggerChars = mergeTriggers(completionTriggerChars,
-									provider.getTriggerCharacters());
+							completionTriggerChars = mergeTriggers(completionTriggerChars,triggerChars);
 						}
 						return CompletableFuture.completedFuture(null);
 					});
-			this.contextInformationLanguageServersFuture = LanguageServers.forDocument(document)
+			contextInformationLanguageServersFuture = LanguageServers.forDocument(document)
 					.withFilter(capabilities -> capabilities.getSignatureHelpProvider() != null) //
 					.collectAll((w, ls) -> {
-						SignatureHelpOptions provider = castNonNull(w.getServerCapabilities()).getSignatureHelpProvider();
+						List<String> triggerChars = castNonNull(w.getServerCapabilities()).getSignatureHelpProvider().getTriggerCharacters();
 						synchronized (contextTriggerCharsSemaphore) {
```

Comment:
I know, not part of your PR, but I think these synchronized blocks can be avoided too. mergeTriggers is a static method so no need for synchronization and contextTriggerChars can be assigned atomicly. I.e. declaring contextTriggerChars volatile should be enough and actually better.

Reply:
Great, I will do that after I have an answer to the other thread in this review.